blog/2023-09-11-weaviate-gorilla-part-1/index.mdx

```diff
@@ -0,0 +1,340 @@
+			“dataType”: [“int”],
+			“description”: “The number of pages in the book.”
+		},
+		{
+			“name”: “isAvailable”,
+			“dataType”: [“boolean”],
+			“description”: “Availability of the book in the library.”
+		},
+		{
+			“name”: “author”,
+			“dataType”: [“Author”],
+			“description”: “The author of the book”
+		}]
+	},
+	{
+		“class”: “Author”,
+		“description”: “An author of books.”,
+		“vectorIndexType”: “hnsw”,
+		“vectorizer”: “text2vec-transformers”,
+		“properties”: [
+		{
+			“name”: “name”,
+			“dataType”: [“text”],
+			“description”: “The name of the author.”
+		},
+		{
+			“name”: “age”,
+			“dataType”: [“text”],
+			“description”: “The age of the author.”
+		}]
+	}
+      ]
+}
+```
+
+Using the Weaviate Gorilla, we can help our fictional friends at GreatReads.com get running as fast as possible! We do this for 50 synthetic schemas across diverse applications in music, cars, AI models, and more!
+
+![Synthetic Schemas](./img/2.png)
+
+Now let’s dive into how we trained the LlaMA-7B model to write Weaviate queries! `WeaviateGraphQLGorilla` learns how to write these queries just like anyone else, by reading the documentation!! Here is an example of the Weaviate API reference of “Hybrid Search”:
+
+![Hybrid Search Docs](./img/3.png)
+
+As mentioned previously, we combine the 50 toy schemas with 46 APIs such as the Hybrid Search example above and use these to prompt `GPT-3.5-turbo` to generate 2,300 examples. This takes 2 hours (128 minutes) to generate 2,300 queries and costs ~$12.00. We then split the data into 1,840 training examples and 460 examples for testing to fine-tune the open-source LlaMA-7B checkpoint with Retrieval-Aware Training.
+
+Retrieval-Aware Training is an exciting direction for the future of Deep Learning in which the inputs to language model optimization are supplemented with retrieved context. We format the training examples by templating natural language commands into an input / prompt template with the retrieved API reference and custom schema.The LLM is then tasked with writing the ground truth GraphQL query from Self-Instruct data creation.
+
+To train this model, we teamed up with Sam Stoelinga and Substratus. The LlaMA-7B model is loaded using HuggingFace’s `AuoTokenizer` and `AuotModelForCausalLM` with float16 weights. The model is then fine-tuned using the amazing `Peft` library from HuggingFace which interfaces the Low-rank Adaptation (LoRA) sparse training strategy. With this setup, training takes roughly 1 hour per epoch on Kuberenetes K8s. Language modeling perplexity loss begins at 1.1 and saturates to ~0.05 in roughly 250 mini-batches as visualized in the following learning curve graph:
+
+![WeaviateGraphQLGorilla Learning Curve](./img/4.png)
+
+Please see this [page](https://www.substratus.ai/) to learn more about Substratus. The training notebook for `WeaviateGraphQLGorilla` can be found [here](https://github.com/weaviate/Gorilla/blob/main/substratus/train-v4.ipynb) as well.
```

Comment:
Great article, thanks for sharing!
I noticed that the link above did not work while I was reading the article [here](https://weaviate.io/blog/weaviate-gorilla-part-1) (private repo?). Will the training notebook be shared soon as well?